clang/lib/Parse/ParseDecl.cpp

```diff
@@ -3774,15 +3775,19 @@ void Parser::ParseDeclarationSpecifiers(DeclSpec &DS,
       isInvalid = DS.SetTypeAltiVecBool(true, Loc, PrevSpec, DiagID, Policy);
       break;
     case tok::kw_pipe:
-      if (!getLangOpts().OpenCL || (getLangOpts().OpenCLVersion < 200 &&
-                                    !getLangOpts().OpenCLCPlusPlus)) {
+      if ((!getLangOpts().OpenCL || (getLangOpts().OpenCLVersion < 200)) &&
+          !getLangOpts().OpenCLCPlusPlus) {
         // OpenCL 2.0 defined this keyword. OpenCL 1.2 and earlier should
         // support the "pipe" word as identifier.
         Tok.getIdentifierInfo()->revertTokenIDToIdentifier();
         goto DoneWithDeclSpec;
       }
       isInvalid = DS.SetTypePipe(true, Loc, PrevSpec, DiagID, Policy);
       break;
+    case tok::kw___pipe:
+      // __pipe is defined only for SYCL kernel language
```

Comment:
Shouldn't we add a check for that?

Reply:
No. If this keyword appears in any other language, clang already will report it be unknown keyword.

Reply:
Yes, but if there is `__pipe` used in an existing library, unrelated to SYCL?
But do we need such a keyword?

Reply:
Fixed.

> But do we need such a keyword?
Will be reworked as attribute.

Reply:
I think the attribute is fine.